### ISSUE TITLE
Desired labels and input_id offset

### ISSUE THREAD

**Issue by pGit1** [tloen/alpaca-lora#171]
The old and updated code produce ```input_ids``` and ```labels``` that are the exact same when examining minibatches generated by the ```trainer``` class. Is this intended? 

Code to reproduce behavior after ```trainer``` has been instantiated with the relevant collate_fn:

```
trainer = transformers.Trainer(
    model=model,
    train_dataset=train_data,
    eval_dataset=val_data,
    args=transformers.TrainingArguments(
        per_device_train_batch_size=micro_batch_size,
        gradient_accumulation_steps=gradient_accumulation_steps,
        warmup_steps=100,
        num_train_epochs=num_epochs,
        learning_rate=learning_rate,
        fp16=True,
        logging_steps=10,
        evaluation_strategy="steps" if val_set_size > 0 else "no",
        save_strategy="steps",
        eval_steps=200 if val_set_size > 0 else None,
        save_steps=200,
        output_dir=output_dir,
        save_total_limit=3,
        load_best_model_at_end=True if val_set_size > 0 else False,
        ddp_find_unused_parameters=False if ddp else None,
        group_by_length=group_by_length,
    ),
    data_collator=transformers.DataCollatorForSeq2Seq(
        tokenizer, pad_to_multiple_of=8, return_tensors="pt", padding=True
    ),
)

d_loader = trainer.get_train_dataloader()
gen = iter(d_loader)
batch = next(gen)
labels = batch['labels']
input_ids = batch['input_ids']
# interrogating labels and input_ids shows they are the same without any offsests
```

Is the lack of offset between ```labels``` and ```input_ids``` the intended behavior? The trainer class is *certainly not* shifting the labels to the right unless this happens after the mini batches are generated in another piece of code. Not sure what the implications are but it would appear the lack of offset forces the model to learn an **identity function** between input and output tokens. :thinking: 

Curious on @tloen, @ikergarcia1996, or anyone else thoughts that understands hugginfface trainer api better than I do.

**Comment by pGit1** [comment]
Closing this but here is what I found in case anyone else is not familiar with huggingface. Boils down to uncommented code in the repo and poor documentation from Transformers tutorials. The answer is here: https://huggingface.co/course/chapter7/6

> ⚠️ Shifting the inputs and labels to align them happens inside the model, so the data collator just copies the inputs to create the labels.

While in this doc: https://huggingface.co/docs/transformers/tasks/language_modeling I was following closely and mimicking said: 

![image](https://user-images.githubusercontent.com/13975114/227737733-821a6235-9da3-4f49-b0cd-8e69db423a05.png)

and made it seem as if the **collate** function was responsible for the label shifting... :angry: . Took me forever to figure this out and hours of reading trainer source code and testing to **no avail**. 

Ultimately found answer here: https://discuss.huggingface.co/t/where-does-the-transformers-do-the-target-text-shifting-in-causal-lm/32408. Th MODELS, NOT THE COLLATE function does the shifting of labels **always**. Causal llm collate functions are just helpers that copy input_ids and makes them labels. For **all** transformer causal llm tasks in huggingface, the forward method of the models implements the shifts, NOT the collate function. smdh...
